Documentation: Clarify packages

youknowriad opened the pull request:
Fixes ARC-167

## Proposed Changes

Calypso has a number of packages, but no one really knows which packages should be usable where, which packages are deprecated... We need to do a good audit of the existing packages as well as providing a documentation page for developers to discover the packages and learn about their current status.

This PR is a first step, it adds a "packages.md" file to the Calypso Guide in the calypso dev docs. It splits the packages into multiple categories, but I'm going to need help to audit all the existing packages. Most of them I don't know anything about. (Feel free to push to the PR)

youknowriad committed:
Documentation: Clarify packages

youknowriad commented on docs/monorepo.md:
```diff
@@ -16,7 +16,20 @@ Directories with packages are:
-Modules should follow our convention for layout:
+## Creating a new package package
+
+### Package guidelines
```
I think it was not clear when and how we should build packages, hopefully this brings some clarity. Let me know if there's more guidelines to add here.

aduth replied:
I think it's good to have these guidelines, though I think I'm still personally confused on what makes sense to live in `client/` vs. `packages/`. A component like [`<JetpackLogo />`](https://github.com/Automattic/wp-calypso/tree/trunk/client/components/jetpack-logo) or a library like [`lib/touch-detect`](https://github.com/Automattic/wp-calypso/tree/trunk/client/lib/touch-detect) have "one clear purpose", but it's not clear to me whether they necessarily make more sense in `packages/` vs. `client/`.

Is there something about `packages/` that we'd want to support or require them being published to NPM? Or that they ought to be applicable to folders outside `client/` (`desktop/`, `apps/`, etc.) or external projects? And if so, should we list those considerations here as well?

youknowriad replied:
> I think it's good to have these guidelines, though I think I'm still personally confused on what makes sense to live in client/ vs. packages/. A component like [<JetpackLogo />](https://github.com/Automattic/wp-calypso/tree/trunk/client/components/jetpack-logo) or a library like [lib/touch-detect](https://github.com/Automattic/wp-calypso/tree/trunk/client/lib/touch-detect) have "one clear purpose", but it's not clear to me whether they necessarily make more sense in packages/ vs. client/.

Good point, It's also not entirely clear to me either and would love thoughts from the others that worked on Calypso more. 

For me, "publishing to npm" and "external projects" both can be solved by marking a package as `"private": true` so I'm not entirely sure why the distinction exists.

aduth commented on docs/guide/packages.md:
I feel like this kind of document would easily fall out of date, and it feels strange to me that we have packages that we wouldn't necessarily recommend, even if that's the current state of things. My level of concern is proportional to the accountability we have attached to "currently auditing all packages". (i.e. if that's actually happening and someone is taking individual ownership over this document, then I'm less concerned)

youknowriad replied:
I agree, I'd love for it to be generated automatically like we do in Gutenberg for the package reference menu. that said, right now (and correct me if I'm wrong), I feel like we're not really in a state where we can do this automatically and recommend everything until we actually officially have a way to mark components as deprecated..., clarify the requirements of the packages in their respective READMEs...

aduth reviewed (approved):
Couple non-blocking thoughts, but LGTM 👍

aduth commented:
In reviewing some of our packages with high number of dependencies, I stumbled upon the [`calypso-package-generator`](https://github.com/Automattic/wp-calypso/tree/trunk/packages/calypso-package-generator) package, which I could see having some value in generating consistent packages with best practices if that's a common problem that it's capable of solving.

Relevant to this pull request, I wonder:

- If the package is something we want to encourage, should we document it in `monorepo.md` alongside our other package guidelines?
- Or if it's no longer useful or valuable to us, would it be safe to remove this package (separately)?

(cc @arthur791004 re: #69883)

tyxla reviewed (commented):
Thanks for working on this one @youknowriad!

Does it make to consider also unifying with the pre-existing (and quite half-baked) packages README file here: https://github.com/Automattic/wp-calypso/blob/trunk/packages/README.md

youknowriad committed:
Move the mono repository documentation to the guide

youknowriad commented on packages/README.md:
```diff
@@ -6,20 +6,4 @@ For projects that can produce independent, binary-like outputs deployed elsewher
-
-`package.json` is linted using ESLint. Run `yarn eslint packages/*/package.json` to validate them.
-
-If you need exceptions to linting rules, add a `./eslintrc.js` file to your package and disable the relevant rules.
```
All of this is a duplication of the monorepo docs.

youknowriad committed:
Add a note on existing packages

youknowriad commented:
Ok So I've changed my mind. I think mainting this list is going to be a pain, and something that I can't commit to at the moment. So for now, I moved the "mono repo" docs to the "Calypso guide" as I believe it's an important subject that everyone should be aware of, and kept the guidelines for new packages. Thoughts?
